### PR TITLE
boolbvt::convert_constant is now independent of bitvector representation

### PR DESCRIPTION
**Pull request by kroening**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [X] My contribution is formatted in line with CODING_STANDARD.md.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by tautschnig — COMMENTED** [review]
What is benefit of this at this time as it is certainly incomplete (just scrolling a few lines up shows another case where the binary string is used directly)? My bigger concern is performance - we will do this particular conversion a lot, and the only saviour might be caching?

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 81e2c269356c802162ad9bd72cc2bb51621f238d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87102225
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 6dfb41278bc3d04d2393de1af33e26e874d4bbcd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87102618

**Comment by kroening** [comment]
Ok, new approach -- this one avoids going through mp_integer.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: e2f2c22526ab691000bf0980226882da7c3cb3f4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87109398

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 9ade5b4223a05b1cdf402e96496dc24ce016d387).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87123885

File: src/util/arith_tools.h [hunk 1]
```diff
@@ -164,4 +164,9 @@ mp_integer power(const mp_integer &base, const mp_integer &exponent);
 void mp_min(mp_integer &a, const mp_integer &b);
 void mp_max(mp_integer &a, const mp_integer &b);
 
+/// get bit with given index from bit vector
+/// \param src: the bitvector representation
+/// \param bit_index: the index (0 is the least significant)
```
Comment (tautschnig) [inline src/util/arith_tools.h]: I guess I don't usually complain about too much documentation, but I'm worried about inconsistencies if we document twice?

File: src/util/simplify_expr.cpp [hunk 2]
```diff
@@ -135,25 +135,22 @@ bool simplify_exprt::simplify_popcount(popcount_exprt &expr)
...
 
-    if(type.id()==ID_signedbv ||
-       type.id()==ID_unsignedbv)
+    if(op_type.id() == ID_signedbv || op_type.id() == ID_unsignedbv)
     {
-      mp_integer value;
-      if(!to_integer(op, value))
-      {
-        std::size_t result;
+      const auto width = to_bitvector_type(op_type).get_width();
```
Comment (tautschnig) [inline src/util/simplify_expr.cpp:142]: How about:
```
const auto value = numeric_cast<mp_integer>(op);
if(!value.has_value())
  return true;

std::size_t result = 0;
while(*value != 0)
{
  ++result;
  *value = bitwise_and(*value, *value - 1);
}

exprt simp_result = from_integer(result, expr.type());
expr.swap(simp_result);

return false;
```

But I don't know which of them is faster in practice, given `bitwise_and` may be expensive.

File: src/util/arith_tools.h (hunk 1 above)
Comment (kroening) [reply src/util/arith_tools.h]: removed

File: src/util/simplify_expr.cpp (hunk 2 above)
Comment (kroening) [reply src/util/simplify_expr.cpp:142]:
```
  unsignedbv_typet u32(32);
  symbol_tablet st;
  const namespacet ns(st);
  simplify_exprt s(ns);
  for(auto i=0; i<1000000; i++)
  {
    popcount_exprt p(from_integer(i, u32), u32);
    s.simplify_popcount(p);
  }
```
5.0s vs. 8.4s in favour of the variant in this PR.

Comment (tautschnig) [reply src/util/simplify_expr.cpp:142]: Ok, thanks!

**Review by tautschnig — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 2313911f2a6839e5a74b4cdabea6aa0e7e9f682f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87130961

**Pull request merged**